site/app/controllers/AuthenticationController.php

```diff
@@ -308,4 +309,200 @@ public function userSelection() {
+        }
+
+        $requirements = json_decode($json, true);
+
+        if ($requirements === null) {
+            return ['all' => true, 'length' => -1];
+        }
+
+        return $requirements;
+    }
+
+    /**
+     * Check if the user ID is valid
+     */
+    public function isAcceptedUserId(string $user_id, string $given_name, string $family_name, string $email): bool {
+        
+        $requirements = $this->getUserIdRequirements();
+
+         // If length is -1, allow any length
+        if ($requirements['length'] !== -1 && strlen($user_id) > $requirements['length']) {
+            return false;
+        }
+
+        if ($requirements['all'] === true) {
+            return true;
+        }
+        elseif ($requirements['require_name'] === true) {
+            $name_requirements = $requirements['name_requirements'];
+            $given_first = $name_requirements['given_first'] === 'true';
+
+            $id_given_name = substr($user_id, ($given_first ? 0 : $name_requirements['family_name']), ($given_first ? $name_requirements['given_name'] : strlen($user_id)));
+            $id_family_name = substr($user_id, ($given_first ? $name_requirements['given_name'] : 0), ($given_first ? strlen($user_id) : $name_requirements['family_name']));
+            $is_given_name = (strtolower($id_given_name) === substr(strtolower($given_name), 0, $name_requirements['given_name']));
+            $is_family_name = (strtolower($id_family_name) === substr(strtolower($family_name), 0, $name_requirements['family_name']));
+            if ($is_family_name && $is_given_name) {
+                return true;
+            }
+            return false;
+        }
+        elseif ($requirements['require_email'] === true) {
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    /**
+     * Returns true if the password is greater than or equal to 12 characters, and has the required characters
+     */
+    public function isGoodPassword(string $password): bool {
+        return strlen($password) >= 12 && $this->checkChars($password);
+    }
+
+    /**
+     * @return array<string>
+     */
+    public function getAcceptedEmails(): array {
+        $json = file_get_contents('/usr/local/submitty/site/accepted_emails.json');
+        if ($json === false) {
+            return ['' => true];
+        }
+
+        // Decode the JSON file
+        $json_data = json_decode($json, true);
+
+        // Check if the JSON was decoded successfully
+        if ($json_data === null) {
+            return ['' => true];
+        }
+        return $json_data;  
+    }
+
+    /**
+     * Checks if the email extension is in the accepted emails JSON file
+     */
+    public function isAcceptedEmail(string $email): bool {
+        $json_data = $this->getAcceptedEmails();
+        // Check if the file was read successfully
+        try {
+            $email_extension = explode('@', $email)[1];
```

Comment:
Multiple @ symbols can actually be in an email address. `"@rpi.edu@"@example.com`, while very strange, is a valid email address. I think the most secure way would be to use some email address parser but I think we can get away as long as you ensure it reads on the last @ symbol.

Reply:
Another comment on this is that you are checking domains but not verifying that the user registering actually owns that address. If we want to protect which domains register, email verification is a must.

Reply:
I changed this to take the last @ symbol, and I am working on adding email verification.